### PR TITLE
🌱 Add failureDomains in MachineDeploymentTopology

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR adds support for defining the `failureDomain` where a MachineDeploymentTopology should deploy machines to

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/5636

/cc @yastij

### CODE REVIEW

File: api/v1beta1/cluster_types.go
Comment (fabriziopandini) [reply api/v1beta1/cluster_types.go:147]: I'll open an issue.
This is another use case where we are putting changes on hold (like for upgrades), so we should make sure that the TopologyReconciled condition is updated accordingly
@ykakarap PTAL

**Comment by vincepri** [comment]
LGTM pending squash

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

File: api/v1beta1/cluster_types.go
Comment (ykakarap) [reply api/v1beta1/cluster_types.go:147]: Noted. Looks like that work will come in a follow up PR.
To help with my understanding, we are looking to update the `FailureDomain` on the MD only after the cluster object is reconciled and `cluster.status.failureDomains` is available? 
OR
Are we planing to hold creating the MD till `cluster.status.failureDomains` is available so that we can validate the provided failureDomain and use that to create the MD.?

Comment (sbueringer) [reply api/v1beta1/cluster_types.go:147]: @ykakarap The second option sounds better to me.

**Pull request merged**
